### DISCUSSION TITLE
controlling table output: image pixel-dimensions vs cells defined in mm

### DISCUSSION THREAD

**Discussion by abubelinha** [py-pdf/fpdf2#990] [Q&A] [answered]
I want to generate a table with controlled dimensions in mm (row height & column width).
In each cell I want to position text and images, which I'll take from internet on the fly.

So these images will have different dimensions in pixels,  and I want to show them in my table at controlled sizes instead of filling in the whole cell. Two reasons for this:
- I might want some images to be smaller than others when they represent different sized objects which I want to show at the same scale (i.e. different value coins, different sized animals, etc).
- I want to put a text footer below the image, so I need some controlled free space in the cell.

  Some comments about my tests:

1. I tried to feed the table with urls of the **same image scaled at different sizes** to compare results. But looks like **fpdf.table() always tries to expand the images to fit the cell (no matter how small the image is or what value you set in `img_fill_width`)**. This is not the expected behavior of `img_fill_width`, is it?.
   So looks like controlling the image size (either original, or downsized with PIL or any other package) does not permit to reduce the final image output size in mm.  
2. There is a way to [control cell padding](https://py-pdf.github.io/fpdf2/Tables.html#setting-cell-padding). That works well but only to limit the maximum image size: again, the smaller images expand up to those limits, showing pixelation: is this the intended behavior? I might want to see them smaller (i.e., different real-sized coins, or properly scaled cat vs mouse: **is a higher padding the only way to reduce image size inside cell** to show a mouse smaller than a cat?).
3. Can I somehow set a **different row height for headers row**? (it does not contain images so it doesn't need to be that high).
4. If I want to **add a footer to my images**, is there a way to do it? The examples I found show an image inside cell, but nothing else (no text above or below image).
   I tried to use [write_html](https://py-pdf.github.io/fpdf2/Tables.html#using-write_html) but that's limited to single line cells, and nesting html tags inside <td></td> is also forbidden (`<br>`, `<p>`) so I couldn't use that trick to show footers for the images.
   I could add intermediate blank rows, and use the new empty cell below each image to put a footer there. But that is an ugly hack so I hope there is an easier way.
5. Regarding [html tables](https://py-pdf.github.io/fpdf2/Tables.html#using-write_html), there it says '_column widths defined as percent of the effective width_'. But **what units are we using to set their '_effective width_'**? 
   In my example I want 130 mm wide tables. But `<table width="130">` rendered a table that was 46 mm width. So I had to hack it like this: `<table border="1" width="{int(130*130/46)}">`
   So it looks like there is a 130/46 = 2.826 ratio between both magnitudes, but I'd like to understand why. 

This is my sample code and my output [table-image-test.pdf](https://github.com/py-pdf/fpdf2/files/13196631/table-image-test.pdf). 
Any corrections/suggestions are very welcome:

```Python
	from fpdf import FPDF
	pdf = FPDF()
	pdf.add_page()
	pdf.set_font("Times", size=12)
	url0 = 'https://upload.wikimedia.org/wikipedia/commons/3/32/Archive-ugent-be-500C7CB6-DFDB-11E5-9D50-9943D43445F2_DS-266_%28cropped%29.jpg'
	url1 = 'https://api.gbif.org/v1/image/cache/fit-in/500x/occurrence/4022958319/media/e41e2174baa3c3898e9ce1fb4bc3b561'
	url2 = 'https://api.gbif.org/v1/image/cache/fit-in/50x/occurrence/4022958319/media/e41e2174baa3c3898e9ce1fb4bc3b561'
	url3 = 'https://api.gbif.org/v1/image/cache/fit-in/10x/occurrence/4022958319/media/e41e2174baa3c3898e9ce1fb4bc3b561'
	TABLE_DATA = (
		('title','image','description','width'),
		('A big image',url0,'Calamintha spp.','1750 px'),
		('A medium image',url1,'Encelia californica','500 px'),
		('A small image',url2,'Encelia californica','50 px'),
		('A dwarf image',url3,'Encelia californica','10 px'),
		)
	print(TABLE_DATA)
	pdf.cell(text="A 130 mm wide table with columns of 35, 30, 45 and 20 mm wide", new_y = 'NEXT', align='C', border=0)
	pdf.ln(); pdf.ln();
	with pdf.table(line_height=33, width=130, col_widths=(35, 30, 45, 20)) as table:
		for data_row in TABLE_DATA:
			row = table.row()
			for datum in data_row:
				if str(datum)[:4]=="http":
					row.cell(img=str(datum), padding=2, img_fill_width=False) 
					# I'd like control these images output size in mm
					# I'd also like to put a footer here (or a header above the image)
				else:
					row.cell(str(datum))
	pdf.add_page()
	pdf.cell(text="A 130 mm wide html-table with columns of 35, 30, 45 and 20 mm wide")
	pdf.ln(); pdf.ln();
	# is there an easier way of setting 35,30,45 and 20 mm column widths for the html-table below? 
	# What's the proper way of setting full html-table width in mm? (130 mm, for example):
	TABLE_DATA = [['<p><img src="'+c+'"></img></p>' if c[:4]=="http" else c for c in r] for r in TABLE_DATA]
	pdf.write_html(f"""<table border="1" width="{int(130*130/46)}"><thead><tr>
	<th width="{int(35/130*100)}%">{TABLE_DATA[0][0]}</th>
	<th width="{int(30/130*100)}%">{TABLE_DATA[0][1]}</th>
	<th width="{int(45/130*100)}%">{TABLE_DATA[0][2]}</th>
	<th width="{int(20/130*100)}%">{TABLE_DATA[0][3]}</th>
	</tr></thead><tbody><tr>
		<td>{'</td><td>'.join(TABLE_DATA[1])}</td>
	</tr><tr>
		<td>{'</td><td>'.join(TABLE_DATA[2])}</td>
	</tr><tr>
		<td>{'</td><td>'.join(TABLE_DATA[3])}</td>
	</tr><tr>
		<td>{'</td><td>'.join(TABLE_DATA[4])}</td>
	</tr></tbody></table>""",
	table_line_separators=True,)
	pdf.output('table-image-test.pdf')
```

**Comment by gmischler** [comment, marked as answer]
<s>Not sure what is going on with the height of your first heading.</s>
Edit: Actually, I think the `line_height=33` is doing that.

Note that `write_html()` always uses `img_fill_width=True`, which explains the different appearance. Both versions look correct in that regard.

Table cells are subject to substantial changes quite soon. If you look at #973 and #339 , then you can get an idea of how it might look in the future. Each table cell will essentialy become a single [text column](https://py-pdf.github.io/fpdf2/TextColumns.html), with all the flexibility in content and formatting that includes. Until then, it is likely that only the most egregious bugs in the table code will be fixed.

> In my example I want 130 mm wide tables. But `<table width="130">` rendered a table that was 46 mm width.

In HTML, a unitless value always defines a size in pixels. Fpdf2 converts that to document units (eg. mm) with the standard PDF resolution of 72 dpi.

**Reply by Lucas-C** [reply]
> In HTML, a unitless value always defines a size in pixels. Fpdf2 converts that to document units (eg. mm) with the standard PDF resolution of 72 dpi.

I would just add that you can defined the unit you want to use in `FPDF` constructor: `FPDF(unit="mm")`

Have you been able to solve your problem @abubelinha?

**Reply by abubelinha** [reply]
> Have you been able to solve your problem @abubelinha?

Sorry I forgot to answer your comment @Lucas-C 
I actually left my script pending because of this @gmischler's comment:
> Table cells are subject to substantial changes quite soon (...) Each table cell will essentialy become a single [text column](https://py-pdf.github.io/fpdf2/TextColumns.html), with all the flexibility in content and formatting that includes

So I am just waiting for those changes to happen because I understand they will made my coding much easier (I am trying to improve a previous reportlab based script which still works ... but as I am liking fpdf2 a lot I want to rewrite it).

I would left the discussion open in the mean time, if you don't mind

**Reply by Lucas-C** [reply]
> I would left the discussion open in the mean time, if you don't mind

That's perfectly fine.

We will try to remember to give you an update, when it will be ready for you to test those incoming changes!

**Comment by abubelinha** [comment]
Hello again.  From time to time (yearly) I get back to my old reportlab script in order to generate a new calendar.
I wonder if there have been any changes during 2024 which can be relevant to the above discussion. 
Thanks a lot and happy new year!

**Reply by Lucas-C** [reply]
Happy new year @abubelinha 🙂

Good idea to ask.

Text columns in table cells have not been implemented yet.

Maybe @gmischler will want to comment a bit more about this.

Else you can find all the changes happing in `fpdf2` there: https://github.com/py-pdf/fpdf2/blob/master/CHANGELOG.md
